src/system/python_loader.cc

```diff
@@ -28,220 +25,177 @@
-            PyObject *format_tb_method = PyObject_GetAttrString(traceback_module, "format_tb");
-            if (format_tb_method) {
-                PyObject * traceback_lines = PyObject_CallFunctionObjArgs(format_tb_method, ptraceback, NULL);
-                if (traceback_lines) {
-                    PyObject * join_str = PyUnicode_FromString("");
-                    PyObject * join = PyObject_GetAttrString(join_str, "join");
-                    PyObject * message = PyObject_CallFunctionObjArgs(join, traceback_lines, NULL);
-                    if (message) {
-                        str_traceback = string(PyUnicode_AsUTF8(PyObject_Str(message)));
-                    }
-                    Py_DECREF(traceback_lines);
-                    Py_DECREF(join_str);
-                    Py_DECREF(join);
-                    Py_DECREF(message);
+void PythonLoader::throw_error(const py::error_already_set &ex) {
+    PyObject *ptype = ex.type().ptr();
+    PyObject *pvalue = ex.value().ptr();
+    PyObject *ptraceback = ex.trace().ptr();
+
+    PyErr_Fetch(&ptype, &pvalue, &ptraceback);
+    string error_description = string(PyUnicode_AsUTF8(PyObject_Str(pvalue)));
+
+    // clear error indicator
+    PyErr_Clear();
+
+    /* See if we can get a full traceback */
+    PyObject *traceback_module = PyImport_ImportModule("traceback");
+
+    string str_traceback;
+    if (traceback_module && ptraceback) {
+        PyObject *format_tb_method = PyObject_GetAttrString(traceback_module, "format_tb");
+        if (format_tb_method) {
+            PyObject * traceback_lines = PyObject_CallFunctionObjArgs(format_tb_method, ptraceback, NULL);
+            if (traceback_lines) {
+                PyObject * join_str = PyUnicode_FromString("");
+                PyObject * join = PyObject_GetAttrString(join_str, "join");
+                PyObject * message = PyObject_CallFunctionObjArgs(join, traceback_lines, NULL);
+                if (message) {
+                    str_traceback = string(PyUnicode_AsUTF8(PyObject_Str(message)));
                 }
+                Py_DECREF(traceback_lines);
+                Py_DECREF(join_str);
+                Py_DECREF(join);
+                Py_DECREF(message);
             }
         }
-        
-        // get value of python's "sys.path"
-        string python_path = PythonLoader::sys_path;
-        replace(python_path.begin(), python_path.end(), ':', '\n');
-        
-        // construct error message
-        string py_message =
-                   "\nType: " + string(PyUnicode_AsUTF8(PyObject_Str(ptype))) + "\n"
-                 + "Message: " + string(PyUnicode_AsUTF8(PyObject_Str(pvalue))) + "\n"
-                 + "Traceback: \n" + str_traceback + "\n"
-                 + "Paths: " + "\n" + python_path + "\n";
-    
-        THROW(ExcPythonError() << EI_PythonMessage( py_message ));
     }
-}
 
+    // get value of python's "sys.path"
+    PythonLoader::get_python_path();
```

Comment:
Should return vector of string (paths) and then join into single string with '\n' separator.